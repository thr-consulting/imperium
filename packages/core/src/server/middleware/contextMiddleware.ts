import debug from 'debug';
import isFunction from 'lodash/isFunction';
import Context from './Context';
import {Connectors, ImperiumRequestHandler, ServerModule} from '../../../types';

const d = debug('imperium.core.server.context');

interface ContextMiddlewareOptions {
	connectors: Connectors,
	modules: ServerModule[],
}

/**
 * Express middleware that creates the context (data models & auth info).
 *
 * Adds .context to the req
 *
 * @param connectors
 * @param modules
 * @returns {function(*, *, *)}
 */
export default function contextMiddleware({connectors, modules}: ContextMiddlewareOptions): ImperiumRequestHandler {
	return (req, res, next) => {
		d('Creating context');

		const context = new Context(connectors); // Create a brand new instance of Context

		if (!modules) throw new Error('Modules are not defined');

		// Loop through each defined module and add it's data models (if any) to the context
		modules.forEach(module => {
			if (module.models && isFunction(module.models)) context.addModule(module.models);
		});

		// Add the context object to the req
		req.context = context;

		next();
	};
}
