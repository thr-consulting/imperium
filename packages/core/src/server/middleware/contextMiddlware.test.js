import contextMiddleware from './contextMiddleware';

const myConnectors = {
	MyConnector: {},
};

function moduleFunc(/* connectors, context */) {
	return {
		MyDataModel: {
			fakeApiFn: () => 'fake api result',
		},
	};
}

describe('context', () => {
	it('should execute the middleware properly', () => {
		const req = {};
		const next = jest.fn();
		contextMiddleware({
			connectors: myConnectors,
			modules: [{models: moduleFunc}],
		})(req, null, next);
		expect(req.context.getModel('MyDataModel')).toMatchSnapshot();
		expect(next.mock.calls.length).toBe(1);
	});
});
