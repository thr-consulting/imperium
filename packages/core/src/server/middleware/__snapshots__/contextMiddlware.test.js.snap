// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`context should execute the middleware properly 1`] = `
Object {
  "fakeApiFn": [Function],
}
`;
